### PR TITLE
Integrate Glint Language Service

### PR DESCRIPTION
**Pull request by cafreeman**
This PR integrates [glint](https://github.com/typed-ember/glint) into Rehearsal via a new `GlintService` and a set of glint-specific plugins. This enables Rehearsal to run on ember or glimmer projects and work with `hbs` and `gjs` files in addition to regular `js` files.

High-level changes:
- Expose Glint via `GlintService` that mirrors `RehearsalService` and can be used with it interchangeably
- Add `GlintCheck` and `GlintFix` plugins for applying codefixes and todos to Glint-specific file types (e.g. `hbs` and `gjs` files)
- Added a filter mechanism for plugins so that you can specify when a plugin should actually run (this allows us to make the glint plugins and rehearsal plugins mutually exclusive)
- Added logic for determining when to use `RehearsalService` vs `GlintService` based on the dependencies present in the root of the project. If we find `ember-source`, `@glimmer/component`, or `@glimmerx/component`, we use Glint.

### CODE REVIEW

**Comment by egorio** [comment]
Looks great!

File: packages/migrate/test/migrate.test.ts
Comment (chadhietala) [inline packages/migrate/test/migrate.test.ts:111]: Is there a reason why we want this here instead of using snapshots?

File: packages/migrate/test/migrate.test.ts
Comment (chadhietala) [inline packages/migrate/test/migrate.test.ts:352]: skip?

File: packages/plugins/package.json
Comment (chadhietala) [inline packages/plugins/package.json:40]: I'm assuming the dependency is just temporary? We likely want this to be a peerDep because rehearsal it self should install this into the developer's project.

File: packages/plugins/src/plugins/glint-fix.plugin.ts
Comment (chadhietala) [inline packages/plugins/src/plugins/glint-fix.plugin.ts:19]: I'm kind of surprised that you don't run into bugs here with looping over the diagnostics and applying changes in the loop. Since we use `diagnostic.range` to compute where we need to splice in the codefix the `range` immediately becomes stale on the next turn of the loop. I think you need to recall `this.getDiagnostics` and re-assign `diagnostics` at the end to ensure you're always getting the new text ranges.

File: packages/plugins/src/plugins/glint-fix.plugin.ts
Comment (chadhietala) [inline packages/plugins/src/plugins/glint-fix.plugin.ts:100]: We should try to consolidate this logic somehow so that we don't have two copies of it in the codebase.

File: packages/migrate/test/migrate.test.ts
Comment (chadhietala) [inline packages/migrate/test/migrate.test.ts:67]: Can we add some cases where we have a `class` based implementation?

**Review by chadhietala — CHANGES_REQUESTED** [review]
Overall looks great. Most of my feedback are clarifying questions / observations but I do think there is a bug in the glint.fix.plugin. The best way to repro it is to have 2 errors in the file and where after applying the fix from first it will be in the text range of the other.

File: packages/migrate/test/migrate.test.ts
Comment (lynchbomb) [reply packages/migrate/test/migrate.test.ts:352]: I looked into this on Saturday. This was the test skipped in the ESM migration. The issue is the way this test is architected. UUID and importer types are being used elsewhere in the monorepo and the types are already available. As such the updated package.json devDep will never be included (since it already is). 

Additionally in debugging (can’t 100% validate but is replicated) if we specify a different module which doesn’t bundle types and we don’t already consume it. TypeScript 4.9 is auto pulling in the correlated @types in the typescript global cache.

File: packages/plugins/package.json
Comment (cafreeman) [reply packages/plugins/package.json:40]: yea I can move this to peer/devDeps before we merge

File: packages/migrate/test/migrate.test.ts
Comment (cafreeman) [reply packages/migrate/test/migrate.test.ts:67]: There's a `with-class` case in this section that does have a class-based component with an inlined `<template>`. Is that what you mean?

File: packages/migrate/test/migrate.test.ts
Comment (lynchbomb) [reply packages/migrate/test/migrate.test.ts:111]: There are a lot of these. @cafreeman if snapshots aren't ideal; what if we just move these to static .output test files?

**Comment by lynchbomb** [comment]
Generally this is awesome. Excellent work! Beyond what is flagged by @chadhietala, merge conflicts and nit around inline expected test output this is good to go (once above is fixed)

File: packages/plugins/src/plugins/glint-fix.plugin.ts
Comment (cafreeman) [reply packages/plugins/src/plugins/glint-fix.plugin.ts]: Updated this to behave like the `diagnostic-fix` plugin (which does what you suggested)

File: packages/migrate/test/migrate.test.ts
Comment (cafreeman) [reply packages/migrate/test/migrate.test.ts:101]: @lynchbomb I'm not opposed to moving to snapshot tests if we really want to do that, but I honestly think having the inlined expected outputs is nice since it makes it *very* easy to see what `migrate` is actually doing.

File: packages/plugins/package.json
Comment (cafreeman) [reply packages/plugins/package.json]: moved

File: packages/plugins/src/plugins/glint-fix.plugin.ts
Comment (egorio) [inline packages/plugins/src/plugins/glint-fix.plugin.ts:71]: Your work looks awesome! 

But I'm curious why does diagnostic plugin was not used directly?
The logic of converting Glint codefixes to TS codefixes cab be hidden inside service.getDiagnostics + logic to applying them can be moved to both services as well, so:

- rehearsal.getCodeFix and gling.getCodeFixes (includes positionToOffset) return the same format
- diagnostic-fix just called geDiagnostic -> getCodeFixes -> applies the same way for both services

**Review by egorio — APPROVED** [review]

File: packages/plugins/src/plugins/glint-fix.plugin.ts
Comment (cafreeman) [reply packages/plugins/src/plugins/glint-fix.plugin.ts:71]: Good question. Given the subtle differences between applying codefixes from Glint vs from Rehearsal it felt more straightforward to just separate them and have a separate plugin for gjs/hbs files rather than trying to conditionally handle everything in one plugin.

**Review by chadhietala — APPROVED** [review]

**Pull request merged**
